File: web/src/types/vendor.ts
// Tipos de Vendedor — espelha backend/internal/dto/vendor.go (camelCase)
// Task 7.1.1: Vendor, CreateVendorRequest, VendorPatch, CommissionRule
// Ref: contracts/api.md §/vendors; data-model.md §vendors; spec §FR-001..FR-005

export type VendorStatus = 'ativo' | 'inativo';

export interface Vendor {
  id: string; // UUID
  name: string; // máx 200 chars (CHK019)
  email: string; // máx 255 chars
  status: VendorStatus;
  anonymizedAt: string | null; // ISO 8601 UTC — LGPD (FR-005)
}

/** @deprecated Use CreateVendorRequest */
export interface VendorCreate {
  name: string;
  email: string;
}

export interface CreateVendorRequest {
  name: string;   // máx 200 chars
  email: string;  // máx 255 chars
  commissionPercentage?: string; // "5.5000" — obrigatório na API; opcional no tipo para compatibilidade
}

export interface VendorPatch {
  name?: string;
  email?: string;
  status?: VendorStatus;
}

/** @deprecated Use VendorPatch */
export interface VendorUpdate {
  name?: string;
  email?: string;
  status?: VendorStatus;
}

// CommissionRule — taxa de comissão de um vendedor em período
// Ref: contracts/api.md §/vendors/{id}/commission-rule; data-model.md §commission_rules
export interface CommissionRule {
  id: string;       // UUID
  vendorId: string;
  percentage: string; // "5.5000" — string para preservar precisão (P-III sem float)
  validFrom: string;  // ISO 8601 date
  validTo: string | null;
  version: number;
}

export interface SetCommissionRuleRequest {
  percentage: string; // "5.5000" — string para preservar precisão
  validFrom: string;  // ISO 8601 date "YYYY-MM-DD"
}

File: backend/internal/dto/vendor.go
// Package dto — Vendor DTOs com json tags camelCase para a API REST.
// Constitution P-III: valores monetários como int64 (centavos); percentuais como string.
package dto

import (
	"financial-dashboard/backend/internal/repository"
)

// VendorResponse é o DTO de saída para um vendedor.
type VendorResponse struct {
	ID           string  `json:"id"`
	Name         string  `json:"name"`
	Email        string  `json:"email"`
	Status       string  `json:"status"`
	AnonymizedAt *string `json:"anonymizedAt,omitempty"`
	CreatedAt    string  `json:"createdAt"`
}

// VendorFromDomain converte um repository.Vendor para VendorResponse.
func VendorFromDomain(v repository.Vendor) VendorResponse {
	resp := VendorResponse{
		ID:        v.ID,
		Name:      v.Name,
		Email:     v.Email,
		Status:    string(v.Status),
		CreatedAt: v.CreatedAt.UTC().Format("2006-01-02T15:04:05Z"),
	}
	if v.AnonymizedAt != nil {
		s := v.AnonymizedAt.UTC().Format("2006-01-02T15:04:05Z")
		resp.AnonymizedAt = &s
	}
	return resp
}

// CreateVendorRequest é o DTO de entrada para criação de vendedor.
type CreateVendorRequest struct {
	Name                 string `json:"name"`
	Email                string `json:"email"`
	CommissionPercentage string `json:"commissionPercentage"` // "8.5000" — string para preservar precisão
}

// UpdateVendorRequest é o DTO de entrada para atualização parcial de vendedor.
type UpdateVendorRequest struct {
	Name       *string `json:"name,omitempty"`
	Email      *string `json:"email,omitempty"`
	Status     *string `json:"status,omitempty"`
	Percentage *string `json:"commissionPercentage,omitempty"`
	Confirm    *bool   `json:"confirm,omitempty"` // para DELETE /vendors/{id} (anonimização LGPD)
}
